cmd/olm/cleanup.go

```diff
@@ -119,3 +127,115 @@ func deleteCatalogSource(crc versioned.Interface, name string) deleteResourceFun
 	}
 }
+
+// cleanupOwnerReferences cleans up inter-namespace and cluster-to-namespace scoped OwnerReferences to ClusterServiceVersions.
+//
+// Cross-namespace and cluster-to-namespace scoped OwnerReferences may cause sibling resources in the owner namespace to be
+// deleted sporadically (see CVE-2019-3884 https://access.redhat.com/security/cve/cve-2019-3884). Older versions of OLM use both types of
+// OwnerReference, and in cases where OLM is updated, they must be removed to prevent erroneous deletion of OLM's self-hosted components.
+func cleanupOwnerReferences(c operatorclient.ClientInterface, crc versioned.Interface) error {
+	listOpts := metav1.ListOptions{}
+	csvs, err := crc.OperatorsV1alpha1().ClusterServiceVersions(metav1.NamespaceAll).List(listOpts)
+	if err != nil {
+		return err
+	}
+
+	uidNamespaces := map[types.UID]string{}
+	for _, csv := range csvs.Items {
+		uidNamespaces[csv.GetUID()] = csv.GetNamespace()
+	}
+	removeBadRefs := crossNamespaceOwnerReferenceRemoval(v1alpha1.ClusterServiceVersionKind, uidNamespaces)
+
+	// Cleanup cross-namespace OwnerReferences on CSVs, ClusterRoles/Bindings, and Roles/Bindings
+	var objs []metav1.Object
+	for _, obj := range csvs.Items {
+		objs = append(objs, &obj)
+	}
+
+	clusterRoles, _ := c.KubernetesInterface().RbacV1().ClusterRoles().List(listOpts)
+	for _, obj := range clusterRoles.Items {
+		objs = append(objs, &obj)
+	}
+
+	clusterRoleBindings, _ := c.KubernetesInterface().RbacV1().ClusterRoleBindings().List(listOpts)
+	for _, obj := range clusterRoleBindings.Items {
+		objs = append(objs, &obj)
+	}
+
+	roles, _ := c.KubernetesInterface().RbacV1().Roles(metav1.NamespaceAll).List(listOpts)
+	for _, obj := range roles.Items {
+		objs = append(objs, &obj)
+	}
+	roleBindings, _ := c.KubernetesInterface().RbacV1().RoleBindings(metav1.NamespaceAll).List(listOpts)
+	for _, obj := range roleBindings.Items {
+		objs = append(objs, &obj)
+	}
+
+	for _, obj := range objs {
+		if !removeBadRefs(obj) {
+			continue
+		}
+
+		update := func() error {
+			// If this is not a type we care about, do nothing
+			return nil
+		}
+		switch v := obj.(type) {
+		case *v1alpha1.ClusterServiceVersion:
+			update = func() error {
+				_, err := crc.OperatorsV1alpha1().ClusterServiceVersions(v.GetNamespace()).Update(v)
+				return err
+			}
+		case *rbacv1.ClusterRole:
+			update = func() error {
+				_, err = c.KubernetesInterface().RbacV1().ClusterRoles().Update(v)
+				return err
+			}
+		case *rbacv1.ClusterRoleBinding:
+			update = func() error {
+				_, err = c.KubernetesInterface().RbacV1().ClusterRoleBindings().Update(v)
+				return err
+			}
+		case *rbacv1.Role:
+			update = func() error {
+				_, err = c.KubernetesInterface().RbacV1().Roles(v.GetNamespace()).Update(v)
+				return err
+			}
+		case *rbacv1.RoleBinding:
+			update = func() error {
+				_, err = c.KubernetesInterface().RbacV1().RoleBindings(v.GetNamespace()).Update(v)
+				return err
+			}
```

Comment:
Why doesn't you make the default case log a warning message and then return? Then you can remove the empty update function assignment above.

Reply:
SGTM. We could also panic since it's probably a programmer error. This already merged, so I may try to sneak this in with another PR at some point.